### ISSUE TITLE
Feeding tensors doesn't stop gradients flowing through

### ISSUE THREAD

**Issue by gaohuazuo** [tensorflow/tensorflow#6903]
Consider this code snippet:

```
a = tf.constant(0)
b = a + 1
c = b * 2
grad, = tf.gradients(c, a)
grad.eval(feed_dict={b: 0})
```

The result is `2`, instead of `0`. To obtain the true gradient, one has to wrap the tensor with an `tf.cond` and feed an additional indicator, which doesn't seem too elegant.

Apologize if I missed something.

**Comment by yaroslavvb** [comment]
It's counter-intuitive, but it would be hard to fix without significantly changing the underlying framework.

Technically `tf.gradients` creates a backprop graph, and the result of running this graph can be different from mathematical gradient. For instance, it can produce `None` values which indicate that given input is not connected to output (as opposed to 0 which means it's connected, but has no numeric effect on it).

Here, the graph you get is below, and when you feed `b` that severs the forward prop chain of computation, but there are still connections from the backprop chain

![gradients](https://cloud.githubusercontent.com/assets/23068/22034207/52df1cb0-dca0-11e6-8d43-69ed83cd5eaa.png)

**Comment by gaohuazuo** [comment]
Is it possible to provide an `is_feed` op that indicates whether a tensor is feed or not? With the help of such op, feed-consistent gradients would be possible.

**Comment by yaroslavvb** [comment]
not sure where one would start for implementing `is_fed` op, maybe someone more familiar with executor internals can comments, @zffchen78 @mrry

**Comment by gaohuazuo** [comment]
Even if we have `is_fed`, it is not easy to eliminate redundant gradient computation. Suppose that `b = f(a); c = g(b)`, the `d_c / d_b` part will still get computed even if `f` has a feed-aware gradient when `b` is fed.

**Comment by yaroslavvb** [comment]
So it seems the issue is that you are requesting TensorFlow to fetch `gradients` node which you know is 0 because you know what feeding part of the graph breaks connection. Meanwhile, TensorFlow has to do some computation before arriving to this conclusion. Could you just feed `0` into the gradients node to save the computation in such cases?

Also, wrapping gradient tensor in `cond` is not sufficient to remove redundant computation, the whole gradients subgraph must be created inside of a `tf.cond` statement

**Comment by gaohuazuo** [comment]
Even when feeding 0 to gradients, `d_b / d_a` still sticks around while being redundant. The only way I can think of is to construct a copy of the graph and simply don't mix gradients with feeding.

BTW, seems that the gradient of `cond` can be optimized harder.

```
a = tf.constant(0., name='a')
p = tf.constant(True, name='p')
b = tf.cond(p, lambda: tf.square(a), lambda: tf.constant(0.), name='b')
c = tf.square(b, name='c')
g, = tf.gradients(c, a, name='g')
```

![graph-run](https://cloud.githubusercontent.com/assets/10446514/22157412/8ca1aa10-df72-11e6-9780-5641eef6f9c3.png)

This is a trace from `g.eval(feed_dict={p: False})`. The `c_grad` subgraph is not actually useful, but anyway get executed (not shaded in tensorboard).

**Comment by spitis** [comment]
If handling this issue in the framework is not practical, is there some way to throw a warning about this whenever a tensor is fed into a node that has an associated gradient computation, or at the very least include it as a warning in the reading data documentation (https://www.tensorflow.org/programmers_guide/reading_data)? This was unexpected behavior to me.

**Comment by girving** [comment]
Agree that this is WAI, for better or worse.
